File: kotlin/src/main/kotlin/345.ReverseVowelsofaString.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/reverse-vowels-of-a-string/


class ReverseVowelsofaString{
  fun reverseVowels(s: String): String {
    val vowels = "aeiouAEIOU"
    val inds = (0 until s.length).filter{ s[it] in vowels }
    var (l, r) = 0 to inds.size-1
    val sl = s.toMutableList()
    while(l < r){
      val t = sl[inds[l]]
      sl[inds[l++]] = sl[inds[r]]
      sl[inds[r--]] = t
    }
    return sl.joinToString("")
  }

  @Test
  fun test1(){
    assertEquals("holle", reverseVowels("hello"))
  }

  @Test
  fun test2(){
    assertEquals("leotcede", reverseVowels("leetcode"))
  }
}

File: python/345_ReverseVowelsofaString.py
from unittest import TestCase
# https://leetcode.com/problems/reverse-vowels-of-a-string/


class ReverseVowelsofaString(TestCase):
    def reverseVowels(self, s):
        """
        :type s: str
        :rtype: str
        """
        vowels = "aeiouAEIOU"
        s = list(s)
        inds = list(filter(lambda i: s[i] in vowels, range(len(s))))
        for i in range(int(len(inds)/2)):
            s[inds[i]], s[inds[-1-i]] = s[inds[-1-i]], s[inds[i]]
        return "".join(s)

    def test1(self):
        self.assertEqual("holle", self.reverseVowels("hello"))

    def test2(self):
        self.assertEqual("leotcede", self.reverseVowels("leetcode"))
